### PR TITLE
FIX: #10, Page uses 100%

### LINKED ISSUE #10: This page uses 100% of one of my CPUs
Seems strange since it just displays text. I uploaded a quick profile from Firefox (the browser in which the page exhibited this behavior) in case that's of any use: https://share.firefox.dev/3dtF1oJ

### PR DESCRIPTION
**Pull request by note89**
issue with function reference beeing new on every rerender
and that function causing a side effect that leads to a rerender.

fix is to memoize the function

 On branch issue-10-page-uses-100-percent
 Changes to be committed:
	modified:   src/pages/index.tsx

### CODE REVIEW

**Comment by note89** [comment]
This has already been deployed but would like a review before settling on this as a accepted solution

File: src/pages/index.tsx [hunk 1]
```diff
@@ -277,10 +277,83 @@ const GithubLink = ({ className }: { className?: string }) => (
...
 ///////////////// END///////////////
 
+// ################
+// #### UTILS #####
+// ################
+
+// Take any function and make memoize it for only one value
+// All consecutive calls will return the initial value (by reference)
+const ___secret = "secrete_value_1238192382384";
+let ___storage: any = ___secret;
```
Comment (jkoppel) [inline src/pages/index.tsx]: Global state?

So this function can only be used once globally?

That seems unstable, and not how I'd interpret the name. Not so much a linguistic antipattern as a plain antipattern.

File: src/pages/index.tsx [hunk 2]
```diff
@@ -277,10 +277,83 @@ const GithubLink = ({ className }: { className?: string }) => (
...
+   * # The Problem
+   *
+   * atomWithHash gets initialized outside of the react component in the official docs
+   * https://jotai.org/docs/utils/atom-with-hash
+   *
+   * While not explicitly mentioning why, if you do not do this then
+   * on each rerender you will have a new function reference and other hooks depending on this one
+   * will rerun aswell, like useAtom below.
+   *
+   * Non of these jotai hooks take any dependency array, so we can be explicit on when to rerun the hook
```
Comment (jkoppel) [inline src/pages/index.tsx]: Do you mean "none" and "can't"?

**Comment by jkoppel** [comment]
I've reviewed the Jotai docs.

Their examples involve invoking `atom` at the top level, outside of any React component.

Seems that would also fix the problem.

What's better about this approach vs. doing it that way?

File: src/pages/index.tsx (hunk 1 above)
Comment (note89) [reply src/pages/index.tsx]: Oops opps. 
Should have been in the closure... 
fixed now

**Comment by note89** [comment]
> I've reviewed the Jotai docs.
> 
> Their examples involve invoking `atom` at the top level, outside of any React component.
> 
> Seems that would also fix the problem.
> 
> What's better about this approach vs. doing it that way?

The thing that is better is that we can use a real existing markdown node as the default value, which is a mandatory field. 
If we call atomWithHash outside react we either need to duplicate the design decisions on how the titles used in the markdown nodes are used for the selected button and the tab slug. 
Or we need to allow for some extra `unset` value. 

This way, we can just use the first element of the list of buttons as the default and the rest works the same. 
Tried to mention this in the comment to the hook.

TLDR; seemed a bit silly to add either an empty string as a default value or write in a correct string for the tab I wanted to be selected on startup, given we might add new antipatterns or change the names of existing ones. 

```
  * 1. Define the atomWithHash outside of react
   *
   * While this works fine we need to supply a default value.
   * Either we use some invalid value like empty string as the default and no tab will be selected
   * (or selected by a useEffect on initial render)
   *
   * Or we duplicate the design name of one of the files to have as the initial value and create hidden coupling
   *
   * Or we could probably at compile time get hold of a file name and inject that into the state.
   * this duplicates a lot of assumptions on how the name is generated from a FileNode though so probably
   * would have to do a bit more then just that to get some value out of it.
   *
   * Gatsby already does a lot of stuff at compile time with the graphQL query so maybe it would be possible
   * to leverage something there.
   ```

All this being said, maybe should just spend a few minutes to see if there is some other package 
more fit for this purpose instead of including a whole state library for this reason. 
On another project, I used https://www.npmjs.com/package/next-usequerystate, so if there was something like that but not next dependent. Or I could write my own thing, probably not that many lines of code.

File: src/pages/index.tsx [hunk 3]
```diff
@@ -277,103 +275,56 @@ const GithubLink = ({ className }: { className?: string }) => (
...
-const once = <Arg extends Array<unknown>, Ret>(
-  fn: (...args: [...Arg]) => Ret
-) => {
-  const ___secret = Symbol();
-  let ___storage: any = ___secret;
-  return (...args: [...Arg]) => {
-    if (___storage === ___secret) {
-      ___storage = fn(...args);
+// A hook that syncs state between the URL and useState
+const useUrlState = (initialValue: Node) => {
```
Comment (jkoppel) [inline src/pages/index.tsx]: Hmmm....is this the pattern of an inappropriately unspecific name?

**Comment by jkoppel** [comment]
Ensure this still fixes the issue, optionally change the underly-specific name, then merge.

File: src/pages/index.tsx (hunk 3 above)
Comment (note89) [reply src/pages/index.tsx]: Yes good point, made it generic.
